### PR TITLE
feat: Separate prod and dev lambda function builds

### PR DESCRIPTION
**Pull request by edwardfoyle**
*Issue #, if available:*
#5103, #4727

*Description of changes:*
Still need to update some tests and write tests but this PR lays the groundwork for distinguishing between "prod" and "dev" lambda function builds.

Refactored all function build / packaging logic to be in the function category plugin

Added a bunch of type definitions

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-category-function/src/commands/function/build.ts [hunk 1]
```diff
@@ -1,18 +1,43 @@
...
-  name: subcommand,
-  run: async context => {
-    const { amplify, parameters } = context;
-    const resourceName = parameters.first;
+/**
+ * To maintain existing behavior, this function builds and then packages lambda functions
+ */
+export const run = async (context: $TSContext) => {
+  const resourceName = context?.input?.subCommands?.[0];
+  const cont =
```
Comment (jhockett) [inline packages/amplify-category-function/src/commands/function/build.ts:15]: nit: is this supposed to be short for 'continue'? If so, I suggest spelling it out

File: packages/amplify-category-function/src/index.ts [hunk 2]
```diff
@@ -148,28 +152,32 @@ export async function initEnv(context) {
...
-    await runtimeManager.invoke({
-      handler: params.handler,
-      event: JSON.stringify(request.event),
-      env: context.amplify.getEnvInfo().envName,
-      runtime: breadcrumbs.functionRuntime,
+// Returns a wrapper around FunctionRuntimeLifecycleManager.invoke() that can be used to invoke the function with only an event
+export async function getInvoker(
+  context: $TSContext,
+  { handler, resourceName, envVars }: InvokerParameters,
+): Promise<({ event: unknown }) => Promise<any>> {
```
Comment (jhockett) [inline packages/amplify-category-function/src/index.ts:159]: nit: `Promise<({ event: unknown }) => Promise<$TSAny>>`

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/packageFunction.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import { pathManager } from 'amplify-cli-core';
+import * as path from 'path';
+import * as fs from 'fs-extra';
+import { Packager } from '../types/packaging-types';
+import { getRuntimeManager } from './functionPluginLoader';
+
+/**
+ * Packages lambda source code and artifacts into a lambda-compatible .zip file
+ */
+export const packageFunction: Packager = async (context, resource) => {
```
Comment (jhockett) [inline packages/amplify-category-function/src/provider-utils/awscloudformation/utils/packageFunction.ts:12]: nit: add types for parameters

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/packageLayer.ts [hunk 4]
```diff
@@ -22,13 +21,14 @@ import { getLayerRuntimes } from './layerRuntimes';
 import crypto from 'crypto';
 import { updateLayerArtifacts } from './storeResources';
 import globby from 'globby';
+import { Packager, PackageRequestMeta } from '../types/packaging-types';
 
-export async function packageLayer(context, resource: Resource) {
+export const packageLayer: Packager = async (context, resource) => {
```
Comment (jhockett) [inline packages/amplify-category-function/src/provider-utils/awscloudformation/utils/packageLayer.ts:26]: nit: parameter types

File: packages/amplify-cli/src/extensions/amplify-helpers/copy-batch.ts [hunk 5]
```diff
@@ -7,15 +7,15 @@ import { JSONUtilities, $TSAny, $TSContext, $TSObject } from 'amplify-cli-core';
  * @param {$TSAny[]}            jobs         - A list of jobs to run.
  * @param {$TSAny}              props        - The props to use for variable replacement.
  * @param {boolean}             force        - Force CF template generation
- * @param {$TSAny[]|$TSObject}  writeParams  - data for the CF template but not parameters file
+ * @param {$TSAny[]|$TSObject}  writeParams  - boolean of whether props should be written to job.paramsFile, or an object that should be written to job.paramsFile
```
Comment (jhockett) [inline packages/amplify-cli/src/extensions/amplify-helpers/copy-batch.ts:10]: nit: update the types in the comment too

File: packages/amplify-util-mock/src/func/index.ts [hunk 6]
```diff
@@ -1,67 +1,53 @@
...
   }
   const { allResources } = await context.amplify.getResourceStatus();
 
+  const event = await resolveEvent(context, resourceName);
   const envVars = hydrateAllEnvVars(allResources, lambdaConfig.environment);
+  context.print.warning('Ensuring latest function changes are built...');
+  await getBuilder(context, resourceName, BuildType.DEV)();
   const invoker = await getInvoker(context, { resourceName, handler: lambdaConfig.handler, envVars });
-  context.print.success('Starting execution...');
+  context.print.warning('Starting execution...');
```
Comment (jhockett) [inline packages/amplify-util-mock/src/func/index.ts:50]: nit: using the warning text for these seems less than ideal. Perhaps another color like`context.print.blue` could be used?

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/packageFunction.ts (hunk 3 above)
Comment (edwardfoyle) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/utils/packageFunction.ts:12]: TS infers the types from the Packager interface

File: packages/amplify-category-function/src/commands/function/build.ts (hunk 1 above)
Comment (edwardfoyle) [reply packages/amplify-category-function/src/commands/function/build.ts:15]: `continue` is a reserved word. Renaming to `confirmContinue`

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/packageLayer.ts (hunk 4 above)
Comment (edwardfoyle) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/utils/packageLayer.ts:26]: same here, types are inferred

File: packages/amplify-util-mock/src/func/index.ts (hunk 6 above)
Comment (edwardfoyle) [reply packages/amplify-util-mock/src/func/index.ts:50]: TIL we have a print.blue method haha. Good call, I'll use that

File: packages/amplify-cli/src/context-extensions.ts [hunk 7]
```diff
@@ -225,7 +225,7 @@ function blue(message: string): void {
   console.log(colors.blue(message));
 }
 
-function fancy(message: string): void {
+function fancy(message?: string): void {
```
Comment (jhockett) [inline packages/amplify-cli/src/context-extensions.ts:228]: Is there a reason only this function has message as optional?

File: packages/amplify-go-function-runtime-provider/src/runtime.ts [hunk 8]
```diff
@@ -35,7 +42,7 @@ export const executeCommand = (
   return output.stdout;
 };
 
-const isBuildStale = (resourceDir: string, lastBuildTimestamp: Date, outDir: string) => {
+const isBuildStale = (resourceDir: string, lastBuildTimeStamp: Date, outDir: string) => {
```
Comment (jhockett) [inline packages/amplify-go-function-runtime-provider/src/runtime.ts:45]: nit: "Timestamp" is one word https://en.wikipedia.org/wiki/Timestamp

File: packages/amplify-cli/src/context-extensions.ts (hunk 7 above)
Comment (edwardfoyle) [reply packages/amplify-cli/src/context-extensions.ts:228]: There was some place where a potentially undefined value was being passed in (unrelated to this PR) and this was the easiest way to get TS to stop complaining

**Review by jhockett — APPROVED** [review]
LGTM

**Pull request merged**